cpp/src/copying/gather_opt.cu

```diff
@@ -0,0 +1,360 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "gather.hpp"
+#include <cudf/copying.hpp>
+#include <cudf/cudf.h>
+#include <utilities/cudf_utils.h>
+#include <utilities/type_dispatcher.hpp>
+#include <cudf/table.hpp>
+#include <string/nvcategory_util.hpp>
+
+#include <algorithm>
+
+#include <table/device_table.cuh>
+
+#include <utilities/column_utils.hpp>
+
+#include <bitmask/bit_mask.cuh>
+#include <cub/cub.cuh>
+
+using bit_mask::bit_mask_t;
+
+namespace impl {
+
+constexpr int warp_size = 32;
+
+template<class CountType, int lane = 0>
+__device__ __inline__ void single_lane_reduce(CountType f, CountType* d_output){
+  
+  static __shared__ gdf_size_type smem[warp_size];
+  
+  int lane_id = (threadIdx.x % warp_size);
+  int warp_id = (threadIdx.x / warp_size);
+  
+  // Assuming lane 0 of each warp holds the value that we want to perform reduction
+  if(lane_id == lane){
+    smem[warp_id] = f;
+  }
+  __syncthreads();
+  
+  if(warp_id == 0){ // Here I am assuming maximum block size is 1024 and 1024 >> 5 = 32
+    f = (lane_id < (blockDim.x / warp_size)) ? smem[lane_id] : 0;
+
+    __shared__ typename cub::WarpReduce<CountType>::TempStorage temp_storage;
+    f = cub::WarpReduce<CountType>(temp_storage).Sum(f);
+    
+    if(lane_id == 0){
+      atomicAdd(d_output, f);
+    }
+  }
+  __syncthreads();
+}
+
+template<bool check_bounds>
+__device__ __inline__ void gather_bitmask_device(
+    const bit_mask_t* __restrict__ source_valid,
+    const gdf_size_type source_row,
+    const gdf_size_type num_source_rows, 
+    bit_mask_t* __restrict__ destination_valid,
+    const gdf_size_type destination_row, 
+    const gdf_size_type num_destination_rows,
+    gdf_size_type* d_count 
+){
+  // instead of setting it to true by default.
+  const uint32_t active_threads = __ballot_sync(0xffffffff, destination_row < num_destination_rows);
+
+  bool source_bit_is_valid = false;
+  if(check_bounds){
+    if(source_row < num_source_rows){
+      source_bit_is_valid = bit_mask::is_valid(source_valid, source_row);
+    }else{
+      // If gather_map does not include this row we should just keep the originla value,
+      source_bit_is_valid = bit_mask::is_valid(destination_valid, destination_row);
+    }
+  }else{
+    source_bit_is_valid = bit_mask::is_valid(source_valid, source_row);
+  }
+  
+  // Use ballot to find all valid bits in this warp and create the output bitmask element
+  const uint32_t valid_warp = __ballot_sync(active_threads, source_bit_is_valid);
+
+  const gdf_index_type valid_index = destination_row / warp_size;
+  // Only one thread writes output
+  if(0 == threadIdx.x % warp_size){
+    destination_valid[valid_index] = valid_warp;
+  }
+  single_lane_reduce(__popc(valid_warp), d_count);
```

Comment:
This will redundantly compute `__popc` on all threads, when only one thread per warp need compute it. Perhaps just pass in `valid_warp` and compute `__popc` inside your `single_lane_reduce` function.

Reply:
If one thread in every warp is doing the popc, there is no additional cost for every thread in the warp to do it. However there is a branching cost to making all but one thread *not* do it.